presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java

```diff
@@ -181,6 +182,19 @@ public HiveWriterFactory(
         this.pageSorter = requireNonNull(pageSorter, "pageSorter is null");
         this.maxSortFilesPerBucket = maxSortFilesPerBucket;
         this.immutablePartitions = immutablePartitions;
+        this.insertExistingPartitionsBehavior = HiveSessionProperties.getInsertExistingPartitionsBehavior(session);
+        if (immutablePartitions) {
+            checkArgument(
+                    insertExistingPartitionsBehavior == InsertExistingPartitionsBehavior.ERROR ||
+                            insertExistingPartitionsBehavior == InsertExistingPartitionsBehavior.OVERWRITE,
+                    format("Unsupported insert existing partition behavior %s with immutable partitions", insertExistingPartitionsBehavior));
+        }
+        else {
+            checkArgument(
+                    insertExistingPartitionsBehavior == InsertExistingPartitionsBehavior.APPEND ||
+                            insertExistingPartitionsBehavior == InsertExistingPartitionsBehavior.OVERWRITE,
+                    format("Unsupported insert existing partition behavior %s with mutable partitions", insertExistingPartitionsBehavior));
```

Comment:
I think it would be reasonable to allow all 3 when immutablePartitions is false.

If possible, this check should be moved to HiveSessionProperty. Check with @dain.